### PR TITLE
IP2Location db4 db7 support

### PR DESCRIPTION
**Pull request by Woeler**
[//]: # ( Please answer the following questions: )

| Q  | A
| --- | ---
| Bug fix? | N/A
| New feature? | **X**
| Related user documentation PR URL | https://github.com/mautic/documentation/pull/192
| Related developer documentation PR URL | N/A
| Issues addressed (#s or URLs) | https://github.com/mautic/mautic/issues/1164
| BC breaks? | N/A
| Deprecations? | N/A

[//]: # ( Note that all new features should have a related user and/or developer documentation PR in their respective repositories. )

[//]: # ( Required: )
#### Description:
Added 2 new IP Lookup options. IP2Location local BIN lookup for DB4BIN and DB7BIN. These databases contain the ISP as well as the organization (domain) behind an IP address. I have also added these fields to the lead view so that one can see these values when they are present.

Note that IP2Location DDB4BIN and DB7BIN are not free databases. Also note that this is my first pull request here. If you have any feedback, please do let me know.

[//]: # ( As applicable: )
#### Steps to reproduce the bug:
N/A

#### Steps to test this PR:
1. Download the free sample of the DB7BIN [here](https://www.ip2location.com/databases/db7-ip-country-region-city-isp-domain)
2. Name the .BIN file "IP-COUNTRY-REGION-CITY-ISP-DOMAIN.BIN" and place it in app/cache/ip_data/
3. In the Backend->Configuration select "IP2Location Local Bin File (DB7BIN)" as your IP Lookup Service
4. Check if data gets added to incoming IP data and see if the fields are present in the Lead view

#### List deprecations along with the new alternative:
N/A

#### List backwards compatibility breaks:
N/A

### CODE REVIEW

File: app/bundles/CoreBundle/IpLookup/IP2LocationBinLookupDB4.php [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+ * @author      Mautic
+ *
+ * @link        http://mautic.org
+ *
+ * @license     GNU/GPLv3 http://www.gnu.org/licenses/gpl-3.0.html
+ */
+
+namespace Beechit\IpBundle\IpLookup;
+
+namespace Mautic\CoreBundle\IpLookup;
```
Comment (escopecz) [inline app/bundles/CoreBundle/IpLookup/IP2LocationBinLookupDB4.php]: I've never seen multiple namespaces defined for one class. What's the reason for it?

**Comment by escopecz** [comment]
Look at the CS issues: https://travis-ci.org/mautic/mautic/jobs/245709832#L596

They can be fixed manually or when you run this command `bin/php-cs-fixer fix -v` (it will run for several minutes)

**Comment by Woeler** [comment]
Thanks John for all your help :)

**Comment by escopecz** [comment]
I'm a bit scared to click this button:
screen shot 2017-06-23 at 10 18 24
Since the file was sent to my email. Can the button be removed if it's not relevant for the service?

**Comment by escopecz** [comment]
It did not fill any information when accessing with this IP address: `5.170.20.147`. What IP address can I use to make some information populated?

**Comment by escopecz** [comment]
I noticed that I was using wrong IP lookup service. I changed it to "IP-COUNTRY-REGION-CITY-ISP-DOMAIN.BIN", visited a page tracked by JS tracking and this appeared in my logs:
```
[2017-06-23 08:39:45] mautic.CRITICAL: Uncaught PHP Exception Doctrine\DBAL\Types\ConversionException: "Serialized array includes null-byte. This cannot be saved as a text. Please check if you not provided object with protected or private members." at app/bundles/CoreBundle/Doctrine/Type/ArrayType.php line 46 {"exception":"[object] (Doctrine\\DBAL\\Types\\ConversionException(code: 0): Serialized array includes null-byte. This cannot be saved as a text. Please check if you not provided object with protected or private members. at app/bundles/CoreBundle/Doctrine/Type/ArrayType.php:46)\n[stacktrace]\n#0 vendor/doctrine/dbal/lib/Doctrine/DBAL/Connection.php(1504): Mautic\\CoreBundle\\Doctrine\\Type\\ArrayType->convertToDatabaseValue(Array, Object(Doctrine\\DBAL\\Platforms\\MySqlPlatform))\n#1 vendor/doctrine/dbal/lib/Doctrine/DBAL/Connection.php(1537): Doctrine\\DBAL\\Connection->getBindingInfo(Array, Object(Mautic\\CoreBundle\\Doctrine\\Type\\ArrayType))\n#2 vendor/doctrine/dbal/lib/Doctrine/DBAL/Connection.php(996): Doctrine\\DBAL\\Connection->resolveParams(Array, Array)\n#3 vendor/doctrine/orm/lib/Doctrine/ORM/Persisters/Entity/BasicEntityPersister.php(480): Doctrine\\DBAL\\Connection->executeUpdate('UPDATE ip_addre...', Array, Array)\n#4 vendor/doctrine/orm/lib/Doctrine/ORM/Persisters/Entity/BasicEntityPersister.php(366): Doctrine\\ORM\\Persisters\\Entity\\BasicEntityPersister->updateTable(Object(Mautic\\CoreBundle\\Entity\\IpAddress), 'ip_addresses', Array, NULL)\n#5 vendor/doctrine/orm/lib/Doctrine/ORM/UnitOfWork.php(1065): Doctrine\\ORM\\Persisters\\Entity\\BasicEntityPersister->update(Object(Mautic\\CoreBundle\\Entity\\IpAddress))\n#6 vendor/doctrine/orm/lib/Doctrine/ORM/UnitOfWork.php(384): Doctrine\\ORM\\UnitOfWork->executeUpdates(Object(Doctrine\\ORM\\Mapping\\ClassMetadata))\n#7 vendor/doctrine/orm/lib/Doctrine/ORM/EntityManager.php(356): Doctrine\\ORM\\UnitOfWork->commit(Object(Mautic\\CoreBundle\\Entity\\IpAddress))\n#8 app/bundles/CoreBundle/Entity/CommonRepository.php(748): Doctrine\\ORM\\EntityManager->flush(Object(Mautic\\CoreBundle\\Entity\\IpAddress))\n#9 app/bundles/CoreBundle/Helper/IpLookupHelper.php(194): Mautic\\CoreBundle\\Entity\\CommonRepository->saveEntity(Object(Mautic\\CoreBundle\\Entity\\IpAddress))\n#10 app/bundles/PageBundle/Model/PageModel.php(433): Mautic\\CoreBundle\\Helper\\IpLookupHelper->getIpAddress()\n#11 app/bundles/PageBundle/Controller/PublicController.php(353): Mautic\\PageBundle\\Model\\PageModel->hitPage(NULL, Object(Symfony\\Component\\HttpFoundation\\Request))\n#12 [internal function]: Mautic\\PageBundle\\Controller\\PublicController->trackingImageAction()\n#13 vendor/symfony/http-kernel/HttpKernel.php(144): call_user_func_array(Array, Array)\n#14 vendor/symfony/http-kernel/HttpKernel.php(64): Symfony\\Component\\HttpKernel\\HttpKernel->handleRaw(Object(Symfony\\Component\\HttpFoundation\\Request), 1)\n#15 vendor/symfony/http-kernel/DependencyInjection/ContainerAwareHttpKernel.php(69): Symfony\\Component\\HttpKernel\\HttpKernel->handle(Object(Symfony\\Component\\HttpFoundation\\Request), 1, true)\n#16 vendor/symfony/http-kernel/Kernel.php(185): Symfony\\Component\\HttpKernel\\DependencyInjection\\ContainerAwareHttpKernel->handle(Object(Symfony\\Component\\HttpFoundation\\Request), 1, true)\n#17 app/AppKernel.php(132): Symfony\\Component\\HttpKernel\\Kernel->handle(Object(Symfony\\Component\\HttpFoundation\\Request), 1, true)\n#18 app/middlewares/CORSMiddleware.php(92): AppKernel->handle(Object(Symfony\\Component\\HttpFoundation\\Request), 1, true)\n#19 app/middlewares/CatchExceptionMiddleware.php(45): Mautic\\Middleware\\CORSMiddleware->handle(Object(Symfony\\Component\\HttpFoundation\\Request), 1, true)\n#20 app/middlewares/Dev/IpRestrictMiddleware.php(66): Mautic\\Middleware\\CatchExceptionMiddleware->handle(Object(Symfony\\Component\\HttpFoundation\\Request), 1, true)\n#21 app/middlewares/VersionCheckMiddleware.php(57): Mautic\\Middleware\\Dev\\IpRestrictMiddleware->handle(Object(Symfony\\Component\\HttpFoundation\\Request), 1, true)\n#22 vendor/stack/builder/src/Stack/StackedHttpKernel.php(23): Mautic\\Middleware\\VersionCheckMiddleware->handle(Object(Symfony\\Component\\HttpFoundation\\Request), 1, true)\n#23 vendor/stack/run/src/Stack/run.php(13): Stack\\StackedHttpKernel->handle(Object(Symfony\\Component\\HttpFoundation\\Request))\n#24 index_dev.php(28): Stack\\run(Object(Stack\\StackedHttpKernel))\n#25 {main}"} 
[2017-06-23 08:39:45] mautic.CRITICAL: Exception thrown when handling an exception (Doctrine\DBAL\Types\ConversionException: Serialized array includes null-byte. This cannot be saved as a text. Please check if you not provided object with protected or private members. at app/bundles/CoreBundle/Doctrine/Type/ArrayType.php line 46) {"exception":"[object] (Doctrine\\DBAL\\Types\\ConversionException(code: 0): Serialized array includes null-byte. This cannot be saved as a text. Please check if you not provided object with protected or private members. at app/bundles/CoreBundle/Doctrine/Type/ArrayType.php:46)\n[stacktrace]\n#0 vendor/symfony/doctrine-bridge/DataCollector/DoctrineDataCollector.php(135): Mautic\\CoreBundle\\Doctrine\\Type\\ArrayType->convertToDatabaseValue(Array, Object(Doctrine\\DBAL\\Platforms\\MySqlPlatform))\n#1 vendor/symfony/doctrine-bridge/DataCollector/DoctrineDataCollector.php(111): Symfony\\Bridge\\Doctrine\\DataCollector\\DoctrineDataCollector->sanitizeQuery('default', Array)\n#2 vendor/symfony/doctrine-bridge/DataCollector/DoctrineDataCollector.php(58): Symfony\\Bridge\\Doctrine\\DataCollector\\DoctrineDataCollector->sanitizeQueries('default', Array)\n#3 vendor/doctrine/doctrine-bundle/DataCollector/DoctrineDataCollector.php(46): Symfony\\Bridge\\Doctrine\\DataCollector\\DoctrineDataCollector->collect(Object(Symfony\\Component\\HttpFoundation\\Request), Object(Symfony\\Component\\HttpFoundation\\Response), NULL)\n#4 vendor/symfony/http-kernel/Profiler/Profiler.php(223): Doctrine\\Bundle\\DoctrineBundle\\DataCollector\\DoctrineDataCollector->collect(Object(Symfony\\Component\\HttpFoundation\\Request), Object(Symfony\\Component\\HttpFoundation\\Response), NULL)\n#5 vendor/symfony/http-kernel/EventListener/ProfilerListener.php(128): Symfony\\Component\\HttpKernel\\Profiler\\Profiler->collect(Object(Symfony\\Component\\HttpFoundation\\Request), Object(Symfony\\Component\\HttpFoundation\\Response), NULL)\n#6 [internal function]: Symfony\\Component\\HttpKernel\\EventListener\\ProfilerListener->onKernelResponse(Object(Symfony\\Component\\HttpKernel\\Event\\FilterResponseEvent), 'kernel.response', Object(Symfony\\Component\\HttpKernel\\Debug\\TraceableEventDispatcher))\n#7 vendor/symfony/event-dispatcher/Debug/WrappedListener.php(61): call_user_func(Array, Object(Symfony\\Component\\HttpKernel\\Event\\FilterResponseEvent), 'kernel.response', Object(Symfony\\Component\\HttpKernel\\Debug\\TraceableEventDispatcher))\n#8 [internal function]: Symfony\\Component\\EventDispatcher\\Debug\\WrappedListener->__invoke(Object(Symfony\\Component\\HttpKernel\\Event\\FilterResponseEvent), 'kernel.response', Object(Symfony\\Component\\EventDispatcher\\ContainerAwareEventDispatcher))\n#9 app/cache/dev/classes.php(2238): call_user_func(Object(Symfony\\Component\\EventDispatcher\\Debug\\WrappedListener), Object(Symfony\\Component\\HttpKernel\\Event\\FilterResponseEvent), 'kernel.response', Object(Symfony\\Component\\EventDispatcher\\ContainerAwareEventDispatcher))\n#10 app/cache/dev/classes.php(2153): Symfony\\Component\\EventDispatcher\\EventDispatcher->doDispatch(Array, 'kernel.response', Object(Symfony\\Component\\HttpKernel\\Event\\FilterResponseEvent))\n#11 vendor/symfony/event-dispatcher/Debug/TraceableEventDispatcher.php(140): Symfony\\Component\\EventDispatcher\\EventDispatcher->dispatch('kernel.response', Object(Symfony\\Component\\HttpKernel\\Event\\FilterResponseEvent))\n#12 vendor/symfony/http-kernel/HttpKernel.php(184): Symfony\\Component\\EventDispatcher\\Debug\\TraceableEventDispatcher->dispatch('kernel.response', Object(Symfony\\Component\\HttpKernel\\Event\\FilterResponseEvent))\n#13 vendor/symfony/http-kernel/HttpKernel.php(166): Symfony\\Component\\HttpKernel\\HttpKernel->filterResponse(Object(Symfony\\Component\\HttpFoundation\\Response), Object(Symfony\\Component\\HttpFoundation\\Request), 2)\n#14 vendor/symfony/http-kernel/HttpKernel.php(64): Symfony\\Component\\HttpKernel\\HttpKernel->handleRaw(Object(Symfony\\Component\\HttpFoundation\\Request), 2)\n#15 vendor/symfony/http-kernel/DependencyInjection/ContainerAwareHttpKernel.php(69): Symfony\\Component\\HttpKernel\\HttpKernel->handle(Object(Symfony\\Component\\HttpFoundation\\Request), 2, false)\n#16 app/bundles/CoreBundle/EventListener/ExceptionListener.php(81): Symfony\\Component\\HttpKernel\\DependencyInjection\\ContainerAwareHttpKernel->handle(Object(Symfony\\Component\\HttpFoundation\\Request), 2, false)\n#17 [internal function]: Mautic\\CoreBundle\\EventListener\\ExceptionListener->onKernelException(Object(Symfony\\Component\\HttpKernel\\Event\\GetResponseForExceptionEvent), 'kernel.exceptio...', Object(Symfony\\Component\\HttpKernel\\Debug\\TraceableEventDispatcher))\n#18 vendor/symfony/event-dispatcher/Debug/WrappedListener.php(61): call_user_func(Array, Object(Symfony\\Component\\HttpKernel\\Event\\GetResponseForExceptionEvent), 'kernel.exceptio...', Object(Symfony\\Component\\HttpKernel\\Debug\\TraceableEventDispatcher))\n#19 [internal function]: Symfony\\Component\\EventDispatcher\\Debug\\WrappedListener->__invoke(Object(Symfony\\Component\\HttpKernel\\Event\\GetResponseForExceptionEvent), 'kernel.exceptio...', Object(Symfony\\Component\\EventDispatcher\\ContainerAwareEventDispatcher))\n#20 app/cache/dev/classes.php(2238): call_user_func(Object(Symfony\\Component\\EventDispatcher\\Debug\\WrappedListener), Object(Symfony\\Component\\HttpKernel\\Event\\GetResponseForExceptionEvent), 'kernel.exceptio...', Object(Symfony\\Component\\EventDispatcher\\ContainerAwareEventDispatcher))\n#21 app/cache/dev/classes.php(2153): Symfony\\Component\\EventDispatcher\\EventDispatcher->doDispatch(Array, 'kernel.exceptio...', Object(Symfony\\Component\\HttpKernel\\Event\\GetResponseForExceptionEvent))\n#22 vendor/symfony/event-dispatcher/Debug/TraceableEventDispatcher.php(140): Symfony\\Component\\EventDispatcher\\EventDispatcher->dispatch('kernel.exceptio...', Object(Symfony\\Component\\HttpKernel\\Event\\GetResponseForExceptionEvent))\n#23 vendor/symfony/http-kernel/HttpKernel.php(221): Symfony\\Component\\EventDispatcher\\Debug\\TraceableEventDispatcher->dispatch('kernel.exceptio...', Object(Symfony\\Component\\HttpKernel\\Event\\GetResponseForExceptionEvent))\n#24 vendor/symfony/http-kernel/HttpKernel.php(75): Symfony\\Component\\HttpKernel\\HttpKernel->handleException(Object(Doctrine\\DBAL\\Types\\ConversionException), Object(Symfony\\Component\\HttpFoundation\\Request), 1)\n#25 vendor/symfony/http-kernel/DependencyInjection/ContainerAwareHttpKernel.php(69): Symfony\\Component\\HttpKernel\\HttpKernel->handle(Object(Symfony\\Component\\HttpFoundation\\Request), 1, true)\n#26 vendor/symfony/http-kernel/Kernel.php(185): Symfony\\Component\\HttpKernel\\DependencyInjection\\ContainerAwareHttpKernel->handle(Object(Symfony\\Component\\HttpFoundation\\Request), 1, true)\n#27 app/AppKernel.php(132): Symfony\\Component\\HttpKernel\\Kernel->handle(Object(Symfony\\Component\\HttpFoundation\\Request), 1, true)\n#28 app/middlewares/CORSMiddleware.php(92): AppKernel->handle(Object(Symfony\\Component\\HttpFoundation\\Request), 1, true)\n#29 app/middlewares/CatchExceptionMiddleware.php(45): Mautic\\Middleware\\CORSMiddleware->handle(Object(Symfony\\Component\\HttpFoundation\\Request), 1, true)\n#30 app/middlewares/Dev/IpRestrictMiddleware.php(66): Mautic\\Middleware\\CatchExceptionMiddleware->handle(Object(Symfony\\Component\\HttpFoundation\\Request), 1, true)\n#31 app/middlewares/VersionCheckMiddleware.php(57): Mautic\\Middleware\\Dev\\IpRestrictMiddleware->handle(Object(Symfony\\Component\\HttpFoundation\\Request), 1, true)\n#32 vendor/stack/builder/src/Stack/StackedHttpKernel.php(23): Mautic\\Middleware\\VersionCheckMiddleware->handle(Object(Symfony\\Component\\HttpFoundation\\Request), 1, true)\n#33 vendor/stack/run/src/Stack/run.php(13): Stack\\StackedHttpKernel->handle(Object(Symfony\\Component\\HttpFoundation\\Request))\n#34 index_dev.php(28): Stack\\run(Object(Stack\\StackedHttpKernel))\n#35 {main}"} 
[2017-06-23 08:39:45] mautic.ERROR: Doctrine\DBAL\Types\ConversionException: Serialized array includes null-byte. This cannot be saved as a text. Please check if you not provided object with protected or private members. - in file app/bundles/CoreBundle/Doctrine/Type/ArrayType.php - at line 46 
```
No contact was created.

**Comment by Woeler** [comment]
You need to select the option for DB7BIN in the configuration. The default IP2Location is DB9BIN.

**Comment by escopecz** [comment]
Uhg, I noticed that in the second comment, but pasted the file name instead of the IP lookup name in the comment. It meant to read:

> I noticed that I was using wrong IP lookup service. I changed it to "IP2Location Local Bin File (DB7BIN)", visited a page...

I was using DB7BIN in the end and that was when I got the errors.

**Comment by Woeler** [comment]
I'll look into it, see if I can replicate.

**Comment by Woeler** [comment]
Hey John,

Can you tell me what you were doing exactly? I have been trying to replicate your error. Had 2 clean installs from my branch and both worked totally fine after installing and selecting the IP lookup service for DB7BIN. Contacts also keep being created.

What database do you use? I'm using MySQL.

**Comment by dongilbert** [comment]
Just a quick FYI - when running `php-cs-fixer`, I've seen some inconsistencies pop up unless you also specify the config file as `php-cs-fixer --config-file=./.php_cs`

**Comment by Woeler** [comment]
@dongilbert Thanks I'll keep that in mind

**Comment by Woeler** [comment]
For some reason travis is now creating a big fuzz about stuff that I haven't even touched.

**Comment by dongilbert** [comment]
No worries on that, I'll get it patched up

**Comment by Woeler** [comment]
It's changing a lot of files on my local install when I run it. I'll not commit that stuff.

**Comment by dongilbert** [comment]
I tested and received no such errors @escopecz. Everything looks good and works well.

**Comment by dongilbert** [comment]
There is an issue though - if you delete your cache (and the ip_data) with a local data-store file selected as your IP data source, you get the following error:

```
InvalidArgumentException: ip2locationlocaldb7bin not registered.
```

**Comment by Woeler** [comment]
I'm pretty sure that always happened for all local databases?

**Comment by Woeler** [comment]
Have a look here https://github.com/mautic/mautic/issues/4313

**Comment by escopecz** [comment]
To finish this PR, it cannot throw an exception if the file is missing. @Woeler is that possible?

**Comment by Woeler** [comment]
Are you asking if I can add an exception if the file is missing?

**Comment by escopecz** [comment]
I hope you are full of energy then :) I was asking if it is possible to catch the exception and log the error message instead of stopping the whole app for it. Like it works now.

**Comment by Woeler** [comment]
I'm a little confused, it has the exact same exception handling as the IP2Location Local BIN lookup right now. The whole lookup procedure is surrounded by a try-catch.

**Comment by escopecz** [comment]
Ah, sorry then. I assumed it's throwing the exception based on [Don's comment[(https://github.com/mautic/mautic/pull/4278#issuecomment-311390811). I'm not sure what is the issue then. @dongilbert can you specify?

**Comment by Woeler** [comment]
Not sure where that exception is thrown, the one mentioned.

**Pull request closed without merging**